components/lib/crash/src/main/java/mozilla/components/lib/crash/service/GleanCrashReporterService.kt

```diff
@@ -0,0 +1,166 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.lib.crash.service
+
+import android.content.Context
+import androidx.annotation.VisibleForTesting
+import mozilla.components.lib.crash.Crash
+import mozilla.components.lib.crash.GleanMetrics.CrashMetrics
+import mozilla.components.service.glean.Glean
+import mozilla.components.support.base.log.logger.Logger
+import mozilla.components.support.ktx.android.content.isMainProcess
+import java.io.File
+
+/**
+ * A [CrashReporterService] implementation for recording metrics with Glean.  The purpose of this
+ * crash reporter is to collect crash count metrics by capturing [Crash.UncaughtExceptionCrash] and
+ * [Crash.NativeCodeCrash] events and record to the respective
+ * [mozilla.components.service.glean.private.CounterMetricType].
+ */
+class GleanCrashReporterService(val context: Context) : CrashReporterService {
+    companion object {
+        // This file is stored in the application's data directory, so it should be located in the
+        // same location as the application.
+        // The format of this file is simple and uses the keys named below, one per line, to record
+        // crashes.  That format allows for multiple crashes to be appended to the file if, for some
+        // reason, the application cannot run and record them.
+        const val CRASH_FILE_NAME = "glean_crash_counts"
+
+        // These keys correspond to the labels found for crashCount metric in metrics.yaml as well
+        // as the persisted crashes in the crash count file (see above comment)
+        const val UNCAUGHT_EXCEPTION_KEY = "uncaught_exception"
+        const val NATIVE_CODE_CRASH_KEY = "native_code_crash"
+    }
+
+    private val logger = Logger("glean/GleanCrashReporterService")
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal val file = File(context.applicationInfo.dataDir, CRASH_FILE_NAME)
+
+    init {
+        // We only want to record things on the main process because that is the only one in which
+        // Glean is properly initialized.  Checking to see if we are on the main process here will
+        // prevent the situation that arises because the integrating app's Application will be
+        // re-created when prompting to report the crash, and Glean is not initialized there since
+        // it's not technically the main process.
+        if (context.isMainProcess() &&
+            Glean.isInitialized() &&
+            checkFileConditions()) {
+
+            // Parse the persisted crashes
+            parseCrashFile()
+
+            // Clear persisted counts by deleting the file
+            file.delete()
+        } else {
+            if (!context.isMainProcess()) {
```

Comment:
nit: I think this could be changed to be slightly more readable. We could re-arrange the `init` like this:

```
init {
   if (!context.isMainProcess()) {
       logger.error("GleanCrashReporterService init off of main process")
       return
   }

   if (!Glean.isInitialized()) {
       logger.error("Glean is not initialized")
       return
   }

  if (!checkFileConditions()) {
     // The function already logs internally about error conditions. 
     return
  }
  
  // Parse the persisted crashes
  parseCrashFile()

  ....
}
```

Reply:
`return` not allowed from `init`...  I suppose I could extract the check to a separate function if you are worried about readability here.

Reply:
> `return` not allowed from `init`...

I ran (pun intended) into this issue too and found that you can wrap things in a `run` block so you can use early returns. See [TabsToolbarFeature](https://github.com/mozilla-mobile/android-components/blob/master/components/feature/tabs/src/main/java/mozilla/components/feature/tabs/toolbar/TabsToolbarFeature.kt#L26).

_carries on driving by_ 😄